neevaco/neeva-android
# Switch to new endpoint for getting incognito tokens

Stop using GraphQL mutations to get incognito tokens.  This means ripping out the pipeline for detecting when the user is performing a search while in incognito, performing a GraphQL call, and then redirecting the load to somewhere else.  Instead, we now perform a httpd call against an endpoint added specifically for clients to create an incognito token.

* When the user first switches to incognito, we fire a request to set up the incognito session token.  This token is saved in memory and not persisted to storage by the app.

* Introduce the `SessionToken`, which is (currently) used for ensuring that a Browser Profile's CookieManager is synchronized with the app.  This includes methods for initializing the `CookieManager` with listeners for the session cookies, as well as methods for sending the cookie back to the `CookieManager`.
  * This same mechanism allows us to read cookies back out of the Incognito profile's CookieManager when the app restarts and then re-cache it in memory.
  * This will eventually be subclassed for keeping track of login and preview tokens.

* Introduce the `IncognitoSessionToken` for keeping track of when the user's incognito cookie changes.  This is fired only when the user triggers the creation of an `IncognitoBrowserWrapper`.  This either loads an existing cookie up from the browser profile or fires a network request to create a new one.

Things that are still missing:
- Like before, we aren't trying to get a new incognito token if the network request fails.

- New unit tests to confirm the behavior.  We've got multiple engineers working on user tokens and `OkHttpClient`s and I want to get this in before the work gets trampled and has to be redone.

## Checklists

### Pointers to help with review
* API changes: https://docs.google.com/document/d/1rRStjSXig6HfbaXPl4cXbJH_jGkRpTWdTIGsY7G9Gx0/edit#

### How was this tested?

### Manual test cases
Confirmed the incognito cookie works:
1. Clear the neeva package's data
2. Open the app
3. Switch to incognito
4. Navigate to neeva.com
5. Confirm you see this image:
![neeva](https://user-images.githubusercontent.com/96092902/190287926-267c7c0d-7f25-465b-a321-84956eefce98.png)

Confirmed the incognito cookie works after restarting:
1. Set the device to enable `Don't keep activities"
2. Open the app
3. Switch to incognito
4. Go back to the user's home screen by hitting the Android home button
5. Navigate to neeva.com
6. Confirm you see this image:
![neeva](https://user-images.githubusercontent.com/96092902/190287926-267c7c0d-7f25-465b-a321-84956eefce98.png)

## Issues addressed
* Work towards https://app.clickup.com/t/3hq9hu6
